pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -664,36 +632,61 @@ private boolean isAckReceiptEnabled(ClientCnx cnx) {
         return ackReceiptEnabled && cnx != null
                 && Commands.peerSupportsAckReceipt(cnx.getRemoteEndpointProtocolVersion());
     }
+}
 
-    private static class LastCumulativeAck {
-        private MessageIdImpl messageId;
-        private BitSetRecyclable bitSetRecyclable;
+@Getter
+class LastCumulativeAck {
 
-        static LastCumulativeAck create(MessageIdImpl messageId, BitSetRecyclable bitSetRecyclable) {
-            LastCumulativeAck op = RECYCLER.get();
-            op.messageId = messageId;
-            op.bitSetRecyclable = bitSetRecyclable;
-            return op;
-        }
+    // It's used as a returned value by `flush()` to avoid creating a new instance each time `flush()` is called
+    public static final FastThreadLocal<LastCumulativeAck> LOCAL_LAST_CUMULATIVE_ACK =
+            new FastThreadLocal<LastCumulativeAck>() {
+
+                @Override
+                protected LastCumulativeAck initialValue() {
+                    return new LastCumulativeAck();
+                }
+            };
+    public static final MessageIdImpl DEFAULT_MESSAGE_ID = (MessageIdImpl) MessageIdImpl.earliest;
+
+    private volatile MessageIdImpl messageId = DEFAULT_MESSAGE_ID;
+    private BitSetRecyclable bitSetRecyclable = null;
+    private boolean flushRequired = false;
 
-        private LastCumulativeAck(Recycler.Handle<LastCumulativeAck> recyclerHandle) {
-            this.recyclerHandle = recyclerHandle;
+    public synchronized void update(final MessageIdImpl messageId, final BitSetRecyclable bitSetRecyclable) {
+        if (this.bitSetRecyclable != null && this.bitSetRecyclable != bitSetRecyclable) {
+            this.bitSetRecyclable.recycle();
         }
+        set(messageId, bitSetRecyclable);
+        flushRequired = true;
+    }
 
-        void recycle() {
+    public synchronized LastCumulativeAck flush() {
+        if (flushRequired) {
+            final LastCumulativeAck localLastCumulativeAck = LOCAL_LAST_CUMULATIVE_ACK.get();
             if (bitSetRecyclable != null) {
-                this.bitSetRecyclable.recycle();
+                localLastCumulativeAck.set(messageId, BitSetRecyclable.valueOf(bitSetRecyclable.toLongArray()));
```

Comment:
Actually not. `BitSetRecyclable#valueOf` doesn't use Netty recycler to create a `BitSetRecyclable` so it doesn't need to recycle. Regarding to the `bitSetRecyclable` in the current instance, it will be recycled next time `update` is called.

Reply:
I'm not sure if it would be proper to add another constructor that accepts a long array to `bitSetRecyclable` to allocate memory from Netty recycler.

```java
    public static BitSetRecyclable create(BitSetRecyclable other) {
        final BitSetRecyclable bitSetRecyclable = create();
        bitSetRecyclable.words = Arrays.copyOf(other.words, other.length());
        bitSetRecyclable.wordsInUse = other.wordsInUse;
        bitSetRecyclable.sizeIsSticky = other.sizeIsSticky;
        return bitSetRecyclable;
    }
```